### PR TITLE
Feature: Create/Delete CloudWatch Alarm Functionality

### PR DESCRIPTION
**Pull request by paulci**
To assist in Autoscaling, newly deployed services with the feature flag `--aws-alarms` enabled will have a corresponding alarm configured in CloudWatch.

### Changes:
* Updates to the CLI options to support minimum required config
* New methods to manage the alarms
* Logic around the feature flag to enable alarm management
* Additional and updated tests

### Testing:
* Unit tests

### Impact:
* None anticipated, although a misconfiguration may cause Ground-Control to fail on launch, or experience difficulties creating/deleting services.

### CODE REVIEW

File: groundcontrol/helpers/cloudwatch.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    """
+    client = get_boto_client(region, 'cloudwatch')
+    cw_alarm_name = '{}-scaling'.format(service_definition['service'])
+    cw_sns_topic = kwargs.get('cw_sns_topic')
+    cw_metric_name = kwargs.get('cw_metric_name')
+    cw_namespace = kwargs.get('cw_namespace')
+    try:
+        client.put_metric_alarm(
+            AlarmName=cw_alarm_name,
+            AlarmDescription='string',
```
Comment (gjtempleton) [inline groundcontrol/helpers/cloudwatch.py]: I'd update this to be a meaningful description.

File: groundcontrol/helpers/cloudwatch.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            ActionsEnabled=True,
+            OKActions=[
+                cw_sns_topic
+            ],
+            AlarmActions=[
+                cw_sns_topic
+            ],
+            InsufficientDataActions=[
+                'string',
```
Comment (gjtempleton) [inline groundcontrol/helpers/cloudwatch.py]: I think this need to be removed, or allowed to be configurable by the user, InsufficientDataActions isn't required to be provided.

File: groundcontrol/helpers/cloudwatch.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            ],
+            MetricName=cw_metric_name,
+            Namespace=cw_namespace,
+            Statistic='SampleCount',
+            ExtendedStatistic='string',
```
Comment (gjtempleton) [inline groundcontrol/helpers/cloudwatch.py]: `ExtendedStatistic` probable needs to be switched to a user providable option (part of kwargs?)

File: groundcontrol/helpers/cloudwatch.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            Dimensions=[
+                {
+                    'Name': 'Service',
+                    'Value': service_definition['service']
+                },
```
Comment (gjtempleton) [inline groundcontrol/helpers/cloudwatch.py]: Provide another dimension in the form of version?

File: groundcontrol/helpers/cloudwatch.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+            ],
+            Period=60,
+            Unit='Seconds',
+            EvaluationPeriods=2,
+            Threshold=8,
+            ComparisonOperator='GreaterThanOrEqualToThreshold'
```
Comment (gjtempleton) [inline groundcontrol/helpers/cloudwatch.py]: I think all 5 options above should become user configurable, probably beyond the scope of the current work however.

File: groundcontrol/helpers/cloudwatch.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    except Exception as e:
+        logger.warning('Unable to create scaling alarm {} - {}'.format(cw_alarm_name, e))
+    return
+
+
+def delete_cloudwatch_alarm(region, cw_alarm_name):
+    client = get_boto_client(region, 'cloudwatch')
+    try:
+        client.delete_alarms(
+            AlarmNames=cw_alarm_name
```
Comment (gjtempleton) [inline groundcontrol/helpers/cloudwatch.py]: This seems counterintuitive to me, I'd update the variable and method names here to make clear it expects a list of alarm names (of 1 or more) or update the way it's being called and perform the wrapping into a list in here.

File: groundcontrol/server.py [hunk 7]
```diff
@@ -12,7 +12,15 @@ def home():
 
 @app.route("/deployment", methods=['POST'])
 def deployment_deploy():
+    region = app.config['AWS_REGION']
     service_definition = flask.request.get_json(force=True)
+    if app.config['CW_CREATE_ALARMS']:
+        create_cloudwatch_alarm(
+            region, service_definition,
+            cw_sns_topic=app.config['CW_SNS_TOPIC'],
```
Comment (gjtempleton) [inline groundcontrol/server.py]: The fact these default to `None` gives a pretty big number of edge cases where someone could set `CW_CREATE_ALARMS` to true and yet fail to provide the required extra options.

File: groundcontrol/server.py [hunk 8]
```diff
@@ -25,10 +33,13 @@ def deployment_query():
 
 @app.route("/deployment", methods=['DELETE'])
 def deregister_service():
-    region = app.config["AWS_REGION"]
+    region = app.config['AWS_REGION']
     payload = flask.request.get_json(force=True)
-    cluster = find_cluster(region, payload["cluster"])
-    flask.request.get_json(force=True)["cluster"] = cluster
+    if app.config['CW_CREATE_ALARMS']:
```
Comment (gjtempleton) [inline groundcontrol/server.py]: This seems like it risks leaving orphaned alarm if a user tests out the feature, deploys a few services, turns the feature off and then tries to delete the services deployed with alarms. Unfortunately I can't see a better option, so probably best to clarify this behaviour in docs/docstring.

File: groundcontrol/helpers/cloudwatch.py (hunk 3 above)
Comment (paulci) [inline groundcontrol/helpers/cloudwatch.py]: I've removed this, I believe it to be outwith the scope of what we want to achieve

File: groundcontrol/helpers/cloudwatch.py (hunk 4 above)
Comment (paulci) [inline groundcontrol/helpers/cloudwatch.py]: The alarm won't be updated on blue/green deploy, so the version would be unreliable

File: groundcontrol/server.py (hunk 8 above)
Comment (paulci) [inline groundcontrol/server.py]: I agree.  My initial thought had been to check for an existing alarm on delete, however not for this iteration.

File: groundcontrol/helpers/cloudwatch.py (hunk 4 above)
Comment (gjtempleton) [inline groundcontrol/helpers/cloudwatch.py]: If that's the case I can't see any way of passing the version through from the Cloudwatch metric, you'll instead be getting metrics for all versions of a service lumped into one metric, meaning one version struggling whilst the other is coping fine would trigger an alarm for the service overall with no context around the struggling version.

**Pull request merged**